File: src/main/java/net/felsstudio/fels/parser/ast/VariableExpression.java
package main.java.net.felsstudio.fels.parser.ast;

import main.java.net.felsstudio.fels.exceptions.VariableDoesNotExistsException;
import main.java.net.felsstudio.fels.lib.Value;
import main.java.net.felsstudio.fels.lib.Variables;

/**
 *
 * @author FelekDevYT,FelsStudio
 */
public final class VariableExpression extends InterruptableNode implements Expression, Accessible {
    
    public final String name;
    
    public VariableExpression(String name) {
        this.name = name;
    }

    @Override
    public Value eval() {
        super.interruptionCheck();
        return get();
    }
    
    @Override
    public Value get() {
        if (!Variables.isExists(name)) throw new VariableDoesNotExistsException(name);
        return Variables.get(name);
    }

    @Override
    public Value set(Value value) {
        Variables.set(name, value);
        return value;
    }
    
    @Override
    public void accept(Visitor visitor) {
        visitor.visit(this);
    }

    @Override
    public <R, T> R accept(ResultVisitor<R, T> visitor, T t) {
        return visitor.visit(this, t);
    }

    @Override
    public String toString() {
        return name;
    }
}

File: src/net/felsstudio/fels/parser/ast/VariableExpression.kt
package net.felsstudio.fels.parser.ast

import net.felsstudio.fels.exceptions.VariableDoesNotExistsException
import net.felsstudio.fels.lib.Value
import net.felsstudio.fels.lib.Variables

/**
 *
 * @author felek
 */
class VariableExpression(@JvmField val name: String) : Expression, Accessible {
    override fun eval(): Value {
        return get()
    }

    override fun get(): Value {
        if (!Variables.isExists(name)) throw VariableDoesNotExistsException(name)
        return Variables.get(name)
    }

    override fun set(value: Value): Value {
        Variables.set(name, value)
        return value
    }

    override fun accept(visitor: Visitor) {
        visitor.visit(this)
    }

    override fun <R, T> accept(visitor: ResultVisitor<R, T>, t: T): R? {
        return visitor.visit(this, t)
    }

    override fun toString(): String {
        return name
    }
}
